### PR TITLE
Move cipher and mode interfaces

### PR DESCRIPTION
**Pull request by reaperhulk**
Move the cipher and mode interfaces. Refs #1508

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -14,6 +14,22 @@ Changelog
   :class:`~cryptography.hazmat.primitives.hashes.HashContext` were moved from
   :mod:`~cryptography.hazmat.primitives.interfaces` to
   :mod:`~cryptography.hazmat.primitives.hashes`.
+* :class:`~cryptography.hazmat.primitives.ciphers.base.CipherContext`,
+  :class:`~cryptography.hazmat.primitives.ciphers.base.AEADCipherContext`,
+  :class:`~cryptography.hazmat.primitives.ciphers.base.AEADEncryptionContext`,
+  :class:`~cryptography.hazmat.primitives.ciphers.base.CipherAlgorithm`, and
+  :class:`~cryptography.hazmat.primitives.ciphers.base.BlockCipherAlgorithm`
+  were moved from :mod:`~cryptography.hazmat.primitives.interfaces` to
+  :mod:`~cryptography.hazmat.primitives.ciphers.base`.
```
Comment (alex) [inline CHANGELOG.rst]: We don't ever document the `.base` portion of the path, they're always exported in `ciphers`

File: docs/hazmat/primitives/mac/cmac.rst [hunk 2]
```diff
@@ -22,7 +22,7 @@ A subset of CMAC with the AES-128 algorithm is described in :rfc:`4493`.
     .. versionadded:: 0.4
 
     CMAC objects take a
-    :class:`~cryptography.hazmat.primitives.interfaces.BlockCipherAlgorithm` provider.
+    :class:`~cryptography.hazmat.primitives.ciphers.base.BlockCipherAlgorithm` provider.
```
Comment (alex) [inline docs/hazmat/primitives/mac/cmac.rst]: Same here.

**Comment by alex** [comment]
So I reviewed this, it looks good, but there's also a whole bunch of subtle crap in it, so I want a second pair of eyes.

**Comment by Ayrx** [comment]
LGTM :+1:

**Pull request merged**